GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerTestCase/DiskLayoutUpgradeTests.cs

```diff
@@ -0,0 +1,131 @@
+            placeholderLines.ShouldContain(x => x == this.PartialFolderPlaceholderString("GVFS"));
+            placeholderLines.ShouldContain(x => x == this.PartialFolderPlaceholderString("GVFS", "GVFS.Common"));
+            placeholderLines.ShouldContain(x => x == this.PartialFolderPlaceholderString("GVFS", "GVFS.Common", "Git"));
+            placeholderLines.ShouldContain(x => x == this.PartialFolderPlaceholderString("GVFS", "GVFS.Tests"));
+        }
+
+        protected void WriteOldPlaceholderListDatabase()
+        {
+            this.fileSystem.WriteAllText(Path.Combine(this.Enlistment.DotGVFSRoot, GVFSHelpers.PlaceholderListFile), PlaceholderListDatabaseContent);
+        }
+
+        protected void PerformIOBeforePlaceholderDatabaseUpgradeTest()
+        {
+            // Create some placeholder data
+            this.fileSystem.ReadAllText(Path.Combine(this.Enlistment.RepoRoot, "Readme.md"));
+            this.fileSystem.ReadAllText(Path.Combine(this.Enlistment.RepoRoot, "Scripts", "RunUnitTests.bat"));
+            this.fileSystem.ReadAllText(Path.Combine(this.Enlistment.RepoRoot, "GVFS", "GVFS.Common", "Git", "GitRefs.cs"));
+
+            // Create a full folder
+            this.fileSystem.CreateDirectory(Path.Combine(this.Enlistment.RepoRoot, "GVFS", "FullFolder"));
+            this.fileSystem.WriteAllText(Path.Combine(this.Enlistment.RepoRoot, "GVFS", "FullFolder", "test.txt"), "Test contents");
+
+            // Create a tombstone
+            this.fileSystem.DeleteDirectory(Path.Combine(this.Enlistment.RepoRoot, "GVFS", "GVFS.Tests", "Properties"));
+
+            string junctionTarget = Path.Combine(this.Enlistment.EnlistmentRoot, "DirJunction");
+            string symLinkTarget = Path.Combine(this.Enlistment.EnlistmentRoot, "DirSymLink");
+            Directory.CreateDirectory(junctionTarget);
+            Directory.CreateDirectory(symLinkTarget);
+
+            string junctionLink = Path.Combine(this.Enlistment.RepoRoot, "DirJunction");
+            string symLink = Path.Combine(this.Enlistment.RepoRoot, "DirLink");
+            ProcessHelper.Run("CMD.exe", "/C mklink /J " + junctionLink + " " + junctionTarget);
+            ProcessHelper.Run("CMD.exe", "/C mklink /D " + symLink + " " + symLinkTarget);
+
+            string target = Path.Combine(this.Enlistment.EnlistmentRoot, "GVFS", "GVFS", "GVFS.UnitTests");
+            string link = Path.Combine(this.Enlistment.RepoRoot, "UnitTests");
+            ProcessHelper.Run("CMD.exe", "/C mklink /J " + link + " " + target);
+            target = Path.Combine(this.Enlistment.EnlistmentRoot, "GVFS", "GVFS", "GVFS.Installer");
+            link = Path.Combine(this.Enlistment.RepoRoot, "Installer");
+            ProcessHelper.Run("CMD.exe", "/C mklink /D " + link + " " + target);
+        }
+
+        protected void PlaceholderDatabaseShouldIncludeCommonLinesForUpgradeTestIO(string[] placeholderLines)

```

Comment:
Do you think the `protected` methods that are only called in `WindowsDiskLayoutUpgradeTests` should be moved to `WindowsDiskLayoutUpgradeTests`?

Since they're using `\` I don't think they would ever apply to other platforms.

(No need to block the PR on this)